feature: Support of ReturnConsumedCapacity in DynamoDBEnhancedClient's TransactWriteItem operation

prateek-vats opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Motivation and Context
Currently the response returned by `TransactWriteItem` operations don't return the Capacity consumed unlike some other requests. This PR introduces that feature
Open issue: https://github.com/aws/aws-sdk-java-v2/issues/4123

## Modifications

- Make TransctWriteItems operation return TransctWriteEnhancedItemResponse which will contain ConsumedCapacity and ItemCollectionMetrics
- Added support for returnCapacityConsumed to the TransctWriteEnhancedItemRequest

## Testing
Added unit tests and functional tests

## Screenshots (if appropriate)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [x] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License
<!--- The SDK is released under the Apache 2.0 license (http://aws.amazon.com/apache2.0/), so any code you submit will be released under that license -->
<!--- For substantial contributions, we may ask you to sign a Contributor License Agreement (http://en.wikipedia.org/wiki/Contributor_License_Agreement) -->
<!--- Put an `x` in the below box if you confirm that this request can be released under the Apache 2 license -->
- [x] I confirm that this pull request can be released under the Apache 2 license

millems commented:
I'll be your PR reviewer! Once https://github.com/aws/aws-sdk-java-v2/pull/5462 is shipped, feel free to update this PR with to fix any matching comments. Hopefully once that one is done, this one will go quick.

prateek-vats commented:
> I'll be your PR reviewer! Once #5462 is shipped, feel free to update this PR with to fix any matching comments. Hopefully once that one is done, this one will go quick.

Sounds great! Thanks @millems

prateek-vats commented:
> I'll be your PR reviewer! Once #5462 is shipped, feel free to update this PR with to fix any matching comments. Hopefully once that one is done, this one will go quick.

@millems It looks like https://github.com/aws/aws-sdk-java-v2/pull/5462 is merged. I've updated this PR as well

millems commented:
I had a few comments, but to help communicate them I just created a commit on top of yours for comparison: https://github.com/aws/aws-sdk-java-v2/commit/4d5c584b5873d47a65688116488bb2ba15370955

I made these changes:
1. A few typos that weren't yours.
2. Removed the type parameters and items field from TransactWriteItemsEnhancedResponse. The type parameter was just being used for the items, and we never actually populated the items.
3. Added a string getter/setter for enums that were missing them.
4. Copy + make immutable the collections in TransactWriteItemsEnhancedResponse.

If you're cool with those changes, you can cherry-pick the commit and I'll give the ship-it tomorrow! Otherwise, let's talk about it.

prateek-vats commented:
> I had a few comments, but to help communicate them I just created a commit on top of yours for comparison: [4d5c584](https://github.com/aws/aws-sdk-java-v2/commit/4d5c584b5873d47a65688116488bb2ba15370955)
> 
> I made these changes:
> 
> 1. A few typos that weren't yours.
> 2. Removed the type parameters and items field from TransactWriteItemsEnhancedResponse. The type parameter was just being used for the items, and we never actually populated the items.
> 3. Added a string getter/setter for enums that were missing them.
> 4. Copy + make immutable the collections in TransactWriteItemsEnhancedResponse.
> 
> If you're cool with those changes, you can cherry-pick the commit and I'll give the ship-it tomorrow! Otherwise, let's talk about it.

Thanks for making these changes. They look great to me. 

> I had a few comments, but to help communicate them I just created a commit on top of yours for comparison: [4d5c584](https://github.com/aws/aws-sdk-java-v2/commit/4d5c584b5873d47a65688116488bb2ba15370955)
> 
> I made these changes:
> 
> 1. A few typos that weren't yours.
> 2. Removed the type parameters and items field from TransactWriteItemsEnhancedResponse. The type parameter was just being used for the items, and we never actually populated the items.
> 3. Added a string getter/setter for enums that were missing them.
> 4. Copy + make immutable the collections in TransactWriteItemsEnhancedResponse.
> 
> If you're cool with those changes, you can cherry-pick the commit and I'll give the ship-it tomorrow! Otherwise, let's talk about it.

Thanks a lot, @millems ,  appreciate your help! I've pulled in your changes

millems commented:
Thanks for your contribution! It will go out with the next regularly-schedule SDK release, some time next week.

prateek-vats commented:
> Thanks for your contribution! It will go out with the next regularly-schedule SDK release, some time next week.

@millems I may have missed running the changelog script. Does that affect the release of these changes?
